app/support/order_row_importer.rb

```diff
@@ -110,23 +118,33 @@ def add_product_to_order
     end
   end
 
+  def purchase_order!(order)
+    if order.validate_order!
+      add_error("Couldn't purchase order") unless order.purchase_without_default_status!
+    else
+      add_error("Couldn't validate order")
+    end
+  end
+
   def backdate_order_details_to_complete!
     @order_details.each do |order_detail|
       order_detail.ordered_at = order_date
       order_detail.backdate_to_complete!(fulfillment_date)
     end
   end
 
-  def chart_string_field
-    @chart_string_field ||= @row[HEADERS[:chart_string]].try(:strip)
+  def field(field)
+    @row[HEADERS[field]].to_s.try(:strip)
   end
 
   def fulfillment_date
-    @fulfillment_date ||= parse_usa_import_date(fulfillment_date_field)
+    @fulfillment_date ||= parse_usa_import_date(field(:fulfillment_date))
   end
 
-  def fulfillment_date_field
-    @fulfillment_date_field ||= @row[HEADERS[:fulfillment_date]].try(:strip)
+  def existing_order
+    return @existing_order if defined?(@existing_order)
+
+    @existing_order = Order.find_by(id: field(:order_number))
```

Comment:
was it intentional to use `find_by`here instead of `find`? do we not care to trigger an exception here if the record isn’t found?

Reply:
The value we're searching for comes straight from user input, so it could easily be not found. In that case, I do want to return `nil` rather than raise an error so I can check for that case (see the first check of `validate_existing_order`) and render an error to present to the user.

What happens when you're bulk uploading is if there are errors we spit out another CSV that should be identical to the uploaded one but with an "Errors" column. This allows the user to fix anything in that file and then upload the modified file.